feat: API access CE system logs.

BenjaminMummery opened the pull request:
[ORQSDK-852]

# The problem

The CE log archive now includes system logs ([schema](https://github.com/zapatacomputing/workflow-driver/blob/92d9ff32189c580fd0a2ff6eec03cc977fd01502/openapi/src/resources/workflow-run-system-logs.yaml#L2)), and there are endpoints available to access them ([slack thread](https://zapatista.slack.com/archives/C02DN7BA04E/p1684237645219149)). However, we don't have a mechanism for getting them in the API.

The result should be accessible on the Python API. We originally wanted to make it part of the data structure returned from WorkflowRun.get_logs() but it might make sense to create a separate method for it, like WorkflowRun.get_system_logs().

# This PR's solution

Implements `WorkflowRun.get_system_logs()` which returns a list of system log strings.

Example:

```python
import orquestra.sdk as sdk

wf_run = sdk.WorkflowRun.by_id("[CE Run ID]")

for log in wf_run.get_system_logs():
    print(log)
```

# Checklist

_Check that this PR satisfies the following items:_

- [X] Tests have been added for new features/changed behavior (if no new features have been added, check the box).
- [x] The [changelog file](CHANGELOG.md) has been updated with a user-readable description of the changes (if the change isn't visible to the user in any way, check the box).
- [X] The PR's title is prefixed with `<feat/fix/chore/imp[rovement]/int[ernal]/docs>[!]:`
- [X] The PR is linked to a JIRA ticket (if there's no suitable ticket, check the box).


[ORQSDK-852]: https://zapatacomputing.atlassian.net/browse/ORQSDK-852?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

BenjaminMummery committed:
feat: API access CE system logs.

[ORQSDK-852]

BenjaminMummery committed:
docs: update changelog.

[ORQSDK-852]

BenjaminMummery committed:
chore: remove outdated todo.

[ORQSDK-852]

BenjaminMummery committed:
chore: remove outdated todo.

[ORQSDK-852]

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-852-api-get-system-logs-from-ce

jamesclark-Zapata commented on src/orquestra/sdk/_base/_api/_wf_run.py:
```diff
@@ -425,6 +425,17 @@ def get_logs(self) -> WorkflowLogs:
         """
         return self._runtime.get_workflow_logs(wf_run_id=self.run_id)
 
+    def get_system_logs(self) -> t.List[str]:
+        """
+        Unstable: this API will change.
+
+        Returns all available system logs printed during execution of this workflow run.
+
+        Returns:
+            A list of log lines printed when running this workflow.
+        """
+        return self._runtime.get_system_logs(wf_run_id=self.run_id)
+
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**suggestion:** ‏Add system logs to the `WorkflowLogs` object returned by `get_logs`

jamesclark-Zapata commented on src/orquestra/sdk/_ray/_ray_logs.py:
```diff
@@ -41,6 +42,72 @@ class WFLog(pydantic.BaseModel):
+    K8S_EVENT = "K8S_EVENT"
+
+
+class K8sEventLogContainerInfoStatus(pydantic.BaseModel):
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** Kubernetes logs are not created by Ray, do they belong here?

jamesclark-Zapata commented on tests/sdk/v2/driver/resp_mocks.py:
```diff
@@ -276,6 +276,17 @@ def make_get_wf_run_logs_response():
+    return "bytes"
+
+
+def make_get_wf_run_system_logs_response_with_content():
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**suggestion:** ‏Can you follow the pattern that Alex mentioned on the call we had? By making this a file that is read in

jamesclark-Zapata commented on tests/sdk/v2/driver/test_ce_runtime.py:
```diff
@@ -1186,6 +1186,58 @@ def test_exception_handling(
+            (_exceptions.InvalidTokenError, exceptions.UnauthorizedError),
+            (_exceptions.ForbiddenError, exceptions.UnauthorizedError),
+            (_exceptions.UnknownHTTPError, _exceptions.UnknownHTTPError),
+            # (
+            #     _exceptions.WorkflowRunLogsNotReadable,
+            #     exceptions.InvalidWorkflowRunLogsError,
+            # ),
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** ‏Commented out code

jamesclark-Zapata commented on tests/sdk/v2/driver/test_client.py:
```diff
@@ -1811,6 +1811,301 @@ def test_unknown_error(
+                    ],
+                )
+
+                assert client.get_system_logs(workflow_run_id) == [
```
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**thought:** ‏This assert compares to a lot of log lines. This make make the test annoying to update in the future.
Do we care specificically about each line? It may be better to test for properties on these logs, like how many are there, do we see some expected lines, etc.

jamesclark-Zapata commented on src/orquestra/sdk/_base/abc.py:
```diff
@@ -145,6 +145,17 @@ def list_workflow_runs(
         """
         raise NotImplementedError()
 
+    # TODO: 852 change to WorkflowLogs?
```
I think we should do this todo now, if not this ticket number is for the ticket this PR is supposed to close.

jamesclark-Zapata commented on src/orquestra/sdk/_base/_driver/_client.py:
```diff
@@ -696,6 +698,84 @@ def get_task_run_logs(self, task_run_id: _models.TaskRunID) -> bytes:
+
+        # Decompress data
+        try:
+            unzipped: bytes = zlib.decompress(resp.content, 16 + zlib.MAX_WBITS)
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** ‏What does `16 + zlib.MAX_WBITS` do here?

jamesclark-Zapata commented on src/orquestra/sdk/_base/_driver/_client.py:
```diff
@@ -696,6 +698,84 @@ def get_task_run_logs(self, task_run_id: _models.TaskRunID) -> bytes:
+        for section in decoded.split("\n"):
+            if len(section) < 1:
+                continue
+            for log in json.loads(section):
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** ‏Can we create a structured type here?

jamesclark-Zapata commented:
OK, in hindsight, best to rebase and work on top of #191

alexjuda commented on src/orquestra/sdk/_base/_driver/_client.py:
```diff
@@ -696,6 +698,84 @@ def get_task_run_logs(self, task_run_id: _models.TaskRunID) -> bytes:
+                        )
+                        logs.append(log[1]["log"])
+                else:
+                    raise NotImplementedError(
```
**issue**: checks like this should be open-ended because the Platform might add more types in the future. Quote from the jira ticket:

* `source_type="RAY_HEAD_NODE"` -> log is and always will be a string
* `source_type="RAY_WORKER_NODE"` -> log is and always will be string
* `source_type="K8S_EVENT"` -> log and always will be a JSON object
* other log data types will be introduced with new values of source_type

alexjuda commented on src/orquestra/sdk/_ray/_ray_logs.py:
```diff
@@ -41,6 +42,72 @@ class WFLog(pydantic.BaseModel):
+# region: CE system logs
+
+
+class SystemLogSourceType(Enum):
```
**nitpick**: please add permalinks to openAPI schema, if possible.

alexjuda commented on src/orquestra/sdk/_base/_driver/_client.py:
```diff
@@ -696,6 +698,84 @@ def get_task_run_logs(self, task_run_id: _models.TaskRunID) -> bytes:
+                        "source type."
+                    ) from e
+
+                if source_type in [
+                    SystemLogSourceType.RAY_HEAD_NODE,
+                    SystemLogSourceType.RAY_WORKER_NODE,
+                ]:
+                    # Logs from RAY_HEAD_NODE or RAY_WORKER_LOAD are strings
+                    logs.append(log[1]["log"])
+                elif source_type in [SystemLogSourceType.K8S_EVENT]:
+                    # Logs from K8S_EVENT are json objects.
+                    try:
+                        interpreted_log = K8sEventLog.parse_obj(log[1]["log"])
```
**suggestion**: I think Pydantic can handle deserializing polymorphic data like this via unions and discriminators, which could eliminate the need for `if-else`s. Have you considered that solution?

BenjaminMummery committed:
feat: API access CE system logs.

[ORQSDK-852]

BenjaminMummery committed:
Separate CE env set up logs from task logs

Separate CE env set up logs from task logs

Add models for logs

Return message objects from driver client

Fix formatting

Split task vs env CE logs

Improve typing with named tuple instead of Annotated

Update CERuntime tests

Add 'other' logs category

Fix formatting

Fix stale ticket number

Move logs binary mock into a file

Add todo

Fix leftovers

Add docstring

Add comment on regexes

Edit changelog

Fix line length

BenjaminMummery committed:
feat: fold system logs in with WorkflowLogs.

BenjaminMummery committed:
feat; ce system logs API cleanup.

[ORQSDK-852]

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-852-api-get-system-logs-from-ce

BenjaminMummery commented on tests/sdk/v2/driver/resp_mocks.py:
```diff
@@ -276,6 +276,17 @@ def make_get_wf_run_logs_response():
+    """
+    As make_get_wf_run_system_logs_response(), but returns bytes that can be decoded.
+    """
+    return b'\x1f\x8b\x08\x00\x00\x00\x00\x00\x00\xff\xec]\xebn\xe46\x96\xce\xefy\x8a\xb3\x02\x16\xe5\xce\x96T\xba\xd4\x1d\xebA\x0c\xdb\x991&\xe3\x18n\'\xc1\xac\xdb\xd0\xb0\xa4SU\x8c%R!)\xbb+A\x03\x8b}\x86Y\xeck\xec;\xec\x9b\xcc\x93,H\xa9.\xbe\xb4\x93N\xcb\xb2l\x97~\xb4\xabU\x14\x8b\xe7\xe3w\x0e/\xe7\x1cJ*\xcc\xec\x84\xcf\xe4\x17\x8fw\xb9\xae\xeb\xf6]\xd7\xfcuo\xff\xf5\x82\x81\xd7\xeb\x7f\xe1u\xb ...
```
I think it's a good idea, done

BenjaminMummery committed:
feat: remove errors I made during rebasing.

[ORQSDK-852]

BenjaminMummery committed:
refactor: move K*s models out of ray logs.

[ORQSDK-852]

BenjaminMummery committed:
fix: remove unnecessary zlib.max_wlib width

[ORQSDK-852]

BenjaminMummery committed:
refactor: use pydatnic discrimination rather than if-else.

[ORQSDK-852]

BenjaminMummery committed:
feat: return message types for system logs.

[ORQSDK-852]

BenjaminMummery committed:
style: correct typehints.

[ORQSDK-852]

jamesclark-Zapata commented on src/orquestra/sdk/_base/_driver/_client.py:
```diff
@@ -694,7 +692,9 @@ def get_task_run_logs(self, task_run_id: _models.TaskRunID) -> bytes:
         return resp.content
 
-    def get_system_logs(self, wf_run_id: _models.WorkflowRunID) -> List[str]:
+    def _get_system_logs(
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** ‏Is there a reason why this is `_` prefixed?

BenjaminMummery replied:
There was, but it's now outdated. I've changed it back to non `_` - prefixed

jamesclark-Zapata commented on tests/sdk/v2/driver/test_client.py:
```diff
@@ -1799,177 +1800,58 @@ def test_logs_decode(
-                ]
+                sys_logs = client._get_system_logs(workflow_run_id)
+
+                assert len(sys_logs)
```
I think this is better, but maybe we can improve it.
Have a look at Alex's tests, by checking the number of messages, the unique files, etc. https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/191/files
I wouldn't want to spend too much longer on it, we can add a ticket to improve later.

jamesclark-Zapata commented on src/orquestra/sdk/_base/_driver/_models.py:
```diff
@@ -431,3 +442,123 @@ class Event(NamedTuple):
+    A pair of ``[timestamp, syslog]``.
+
+    Based on:
+    https://github.com/zapatacomputing/workflow-driver/blob/972aaa3ca75780a52d01872bc294be419a761209/openapi/src/resources/workflow-run-logs.yaml#L18
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** ‏This should be the system logs defintion.

BenjaminMummery replied:
It should indeed, updated :)

jamesclark-Zapata commented on src/orquestra/sdk/_base/_driver/_models.py:
```diff
@@ -431,3 +442,123 @@ class Event(NamedTuple):
+
+    RAY_HEAD_NODE = "RAY_HEAD_NODE"
+    RAY_WORKER_NODE = "RAY_WORKER_NODE"
+    K8S_EVENT = "K8S_EVENT"
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** ‏Should we include an "unknown" source type that uses the log as a string?

BenjaminMummery replied:
Good idea, I've implemented an UNKNOWN value for the enum with a corresponding event log :)

BenjaminMummery committed:
docs: update schema permalink.

[ORQSDK-852]

BenjaminMummery committed:
test: improved tests for system logs.

[ORQSDK-852]

BenjaminMummery committed:
chore: remove unnecessary _ prefix.

[ORQSDK-852]

alexjuda commented on src/orquestra/sdk/_base/_driver/_models.py:
```diff
@@ -431,3 +442,123 @@ class Event(NamedTuple):
+    message: Optional[str]
+
+
+class K8sEventLogContent(pydantic.BaseModel):
```
**issue**: I couldn't find an API spec for these fields, it's only specified as an [example](https://github.com/zapatacomputing/workflow-driver/blob/2185489dd31fee86e3953bddb37c1fd3754dcde8/openapi/src/resources/workflow-run-system-logs.yaml#L50). We shouldn't rely on behavior outside of the API contract.

It's also a lot of boilerplate to parse and interpret content of these JSONs. The system logs are meant to be noisy, a low-risk, last-effort place to look for errors. I'd suggest we don't even attempt to interpret them and return to the user whatever Platform sent us.

Code is liability; the less code we have to maintain the better!

alexjuda replied:
> I'd suggest we don't even attempt to interpret them and return to the user whatever Platform sent us.

By "them" I mean the content of a `K8S_EVENT` JSON object.

alexjuda commented on src/orquestra/sdk/_base/_driver/_models.py:
```diff
@@ -431,3 +442,123 @@ class Event(NamedTuple):
+
+    tag: str
+
+    log: K8sEventLogContent
```
```suggestion
    log: Dict
    """
    The keys in this dictionary are determined by Kubernetes.
    """
```

BenjaminMummery committed:
feat: handle unknown source types.

[ORQSDK-852]

BenjaminMummery committed:
feat: remove parsing of k8s logs - leave it as a dict.

[ORQSDK-852]

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-852-api-get-system-logs-from-ce

BenjaminMummery committed:
fix: remove unused conditional in Syslog

[ORQSDK-852]

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-852-api-get-system-logs-from-ce

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-852-api-get-system-logs-from-ce
